### PR TITLE
JENKINS-59301 a solutions - Crowd 2 plugin not compatible with JDK11

### PR DESCRIPTION
**Pull request by Rifffi**
A classloader hack was implemented before and after the rest method call. That is
only for JAVA >= 11
see https://www.jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/

update some dependencies
changed:
crowd rest client 4.0.2
commons-lang3 3.8

add minimum jenkins version to 2.163 because jaxb plugin need this

### CODE REVIEW

**Comment by Rifffi** [comment]
Used the test case from https://github.com/jenkinsci/crowd2-plugin/pull/37.

**Comment by KalleOlaviNiemitalo** [comment]
Is this hack compatible with <https://github.com/jenkinsci/jenkins/pull/5698>?

**Comment by Rifffi** [comment]
We are currently using Jenkins version 2.312 and the bouncycastle API 2.23, and this crowd2 version. Maybe I will be able to update and test to more recent versions in the near future.

**Comment by Rifffi** [comment]
The test was successful with Verison 2.316 and bouncycastle API 2.25. Another test with the option hudson.ClassicPluginStrategy.useAntClassLoader=false was also successful.

**Comment by mahernandezp** [comment]
I've tested this PR in jenkins/jenkins:2.303.3 docker image (latest LTS). It works!.
Thank you

**Comment by ricksbrown** [comment]
Looks like the maintainer has moved on. The plugin is "up for adoption": https://plugins.jenkins.io/crowd2/

**Comment by ricksbrown** [comment]
This change does indeed work on 2.303.3.

Did not seem to work using [IBM's version of Java 11](https://developer.ibm.com/languages/java/semeru-runtimes/downloads) - Semeru. Could have been something else I did, but switched to OpenJDK and it works.

**Comment by preddymm** [comment]
@mahernandezp Could you share what changes you have made to work on jenkins/jenkins:2.303.3. I tried the same way what it was shown in https://github.com/jenkinsci/crowd2-plugin/pull/40.

**Comment by Rifffi** [comment]
Don't change anything on the pom.xml. Than run maven with the goals _clean install_
You can also try to disable checks and tests _clean install -Denforcer.fail=false -Dspotbugs.failOnError=false -DskipTests_

<jenkins.version>2.163</jenkins.version> stands for the most minimal version

**Comment by Rifffi** [comment]
> This change does indeed work on 2.303.3.
> 
> Did not seem to work using [IBM's version of Java 11](https://developer.ibm.com/languages/java/semeru-runtimes/downloads) - Semeru. Could have been something else I did, but switched to OpenJDK and it works.

Could be a bug in the rest of the crowd library. The detection that it is java 11 works.

**Comment by Rifffi** [comment]
> @mahernandezp Could you share what changes you have made to work on jenkins/jenkins:2.303.3. I tried the same way what it was shown in #40. It doesn't worked for me for 2.303.3.
> 
> When I do the mvn package by changing the Jenkins version, It is giving me the following error.
> 
> [WARNING] Rule 4: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message: Failed while enforcing RequireUpperBoundDeps. The error(s) are [ Require upper bound dependencies error for com.github.spotbugs:spotbugs-annotations:3.1.12 paths to dependency are: +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-com.github.spotbugs:spotbugs-annotations:3.1.12 and +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.jenkins-ci.main:jenkins-core:2.303.3 +-org.kohsuke.stapler:stapler:1563.v3da2d02f9572 +-com.github.spotbugs:spotbugs-annotations:3.1.12 (managed) <-- com.github.spotbugs:spotbugs-annotations:4.2.3 , Require upper bound dependencies error for org.slf4j:jcl-over-slf4j:1.7.25 paths to dependency are: +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.slf4j:jcl-over-slf4j:1.7.25 and +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.jenkins-ci.main:jenkins-core:2.303.3 +-org.slf4j:jcl-over-slf4j:1.7.31 , Require upper bound dependencies error for org.slf4j:log4j-over-slf4j:1.7.25 paths to dependency are: +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.slf4j:log4j-over-slf4j:1.7.25 and +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.jenkins-ci.main:jenkins-core:2.303.3 +-org.slf4j:log4j-over-slf4j:1.7.31 , Require upper bound dependencies error for org.kohsuke:access-modifier-annotation:1.16 paths to dependency are: +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.kohsuke:access-modifier-annotation:1.16 and +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.jenkins-ci.main:jenkins-core:2.303.3 +-org.kohsuke:access-modifier-annotation:1.21 , Require upper bound dependencies error for org.slf4j:slf4j-jdk14:1.7.25 paths to dependency are: +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.slf4j:slf4j-jdk14:1.7.25 and +-org.jenkins-ci.plugins:crowd2:2.0.3-SNAPSHOT +-org.jenkins-ci.main:jenkins-war:2.303.3 +-org.slf4j:slf4j-jdk14:1.7.31 ]

Don't change anything on the pom.xml. Than run maven with the goals _clean install_
You can also try to disable checks and tests _clean install -Denforcer.fail=false -Dspotbugs.failOnError=false -DskipTests_

<jenkins.version>2.163</jenkins.version> stands for the most minimal version

**Comment by irrationalism** [comment]
FYI: We have it running on the newest jenkins lts version without any problems.

**Comment by preddymm** [comment]
@Rifffi @mahernandezp @irrationalism 
It worked for me. Thank you so much!
I have a question on https://issues.jenkins.io/browse/JENKINS-64186
Any idea on this? I have to check disable remember me to work this.

Thanks

**Comment by preddymm** [comment]
@Rifffi @irrationalism @mahernandezp @KalleOlaviNiemitalo @ricksbrown 

I have a question on https://issues.jenkins.io/browse/JENKINS-64186
Any idea on this? I have to check disable remember me to work this.

Please do let me know your ideas on this

File: pom.xml [hunk 1]
```diff
@@ -46,8 +46,8 @@
     </contributors>
 
     <properties>
-        <jenkins.version>2.89.3</jenkins.version>
-        <crowd-integration-client-rest.version>2.8.8</crowd-integration-client-rest.version>
+        <jenkins.version>2.163</jenkins.version>
+        <crowd-integration-client-rest.version>4.0.2</crowd-integration-client-rest.version>
```
Comment (DuMaM) [inline pom.xml:50]: I would left current version of crowd for now. This change will only provide a JAVA11 compatibility.
Bumping rest client version would be a breaking change.

Comment (DuMaM) [inline pom.xml:50]:
```suggestion
        <crowd-integration-client-rest.version>2.8.8</crowd-integration-client-rest.version>
```

**Review by DuMaM — CHANGES_REQUESTED** [review]

**Comment by DuMaM** [comment]
I need to add some changes required by org first.
https://github.com/jenkinsci/crowd2-plugin/pull/42
Then I will be able to integrate this PR.

**Review by DuMaM — APPROVED** [review]

**Pull request merged**